spark/src/main/scala/org/apache/spark/sql/delta/managedcommit/ManagedCommitUtils.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.managedcommit
+
+import org.apache.spark.sql.delta.{DeltaLog, Snapshot}
+import org.apache.spark.sql.delta.util.FileNames
+import org.apache.spark.sql.delta.util.FileNames.{CompactedDeltaFile, DeltaFile, UnbackfilledDeltaFile}
+import org.apache.hadoop.fs.FileStatus
+
+object ManagedCommitUtils {
+
+  /**
+   * Returns an iterator of commit files starting from startVersion.
+   * If the iterator is consumed beyond what the file system listing shows, this method do a
+   * deltaLog.update() to find the latest version and returns listing results upto that version.
+   *
+   * @return an iterator of (file status, version) pair corresponding to commit files
+   */
+  def commitFilesIterator(
+      deltaLog: DeltaLog,
+      startVersion: Long): Iterator[(FileStatus, Long)] = {
+
+    def listDeltas(startVersion: Long, endVersion: Option[Long]): Iterator[(FileStatus, Long)] = {
+      deltaLog
+        .listFrom(startVersion)
+        .collect { case DeltaFile(fileStatus, version) => (fileStatus, version) }
+        .takeWhile { case (_, version) => endVersion.forall(version <= _) }
+    }
+
+    var maxVersionSeen = startVersion - 1
+    val listedDeltas = listDeltas(startVersion, endVersion = None).filter { case (_, version) =>
+      maxVersionSeen = math.max(maxVersionSeen, version)
+      true
+    }
+
+    def tailFromSnapshot(): Iterator[(FileStatus, Long)] = {
+      val currentSnapshotInDeltaLog = deltaLog.unsafeVolatileSnapshot
+      if (currentSnapshotInDeltaLog.version == maxVersionSeen &&
+          currentSnapshotInDeltaLog.tableCommitStoreOpt.isEmpty) {
```

Comment:
To make sure I understand: If the volatile snapshot version is _smaller_ than what the listing found, then we don't know whether the table might have managed commits or not, because the feature could have been enabled by one of the later commits we don't yet have a snapshot for?

What if the unsafe snapshot version is _larger_ than what the listing returned tho? In that case it's probably ok to trust the listing result as long as the snapshot doesn't have MC enabled... but I guess it's also a _really_ small window where that could happen, so we probably don't care to optimize for it?

Reply:
> If the volatile snapshot version is smaller than what the listing found, then we don't know whether the table might have managed commits or not, because the feature could have been enabled by one of the later commits we don't yet have a snapshot for?

correct.

> What if the unsafe snapshot version is larger than what the listing returned tho? In that case it's probably ok to trust the listing result as long as the snapshot doesn't have MC enabled... but I guess it's also a really small window where that could happen, so we probably don't care to optimize for it?

If unsafeVolatileSnapshot is larger than what listing returned, then:
1. [common case] Either MC is enabled and commits b/w listing-end and unsafeVolatileSnapshot are not backfilled yet?
2. Or MC just got disabled and backfill has not finished yet. uncommon case - we don't want to optimize this.

Reply:
Yes, that's about what I was thinking as well. Not safe to optimize in general, and even the somewhat safe case is uncommon so meh.

Reply:
There is this another case:
LIST returns: 1, 2, 3, 4, 5, 6, volatile snapshot is v10 with no commit store
=> the snapshot may have updated after our listing completed (very narrow time window, but maybe the thread got pre-empted by OS or something)
=> by linearizability arg, we can safely claim the latest version as-of our operation is v6, and ignore the snapshot